californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
> 				/* keep track of the starting request. Currently only used with blockwise transfer */
> 				Request request = exchange.getRequest();
> 				if (request != originRequest && null != request &&  null != request.getToken()
> 						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
> 					// remove starting request also
> 					originRequest = request;
> 					idByToken = KeyToken.fromOutboundMessage(originRequest);
> 					exchangeStore.remove(idByToken, exchange);
> 				}
> 

I thought you was complainig about the blockwise specific stuff in the general ExchangeObserver. Therefore I tried to check, if moving this is posssible. I think it woulb be too complicated to move these line into a special BlockwiseExchangeObserver.